shards/shards.go

```diff
@@ -535,9 +543,34 @@ func (ss *shardedSearcher) streamSearch(ctx context.Context, proc *process, q qu
 					metricSearchMatchCountTotal.Add(float64(sr.Stats.MatchCount))
 					metricSearchNgramMatchesTotal.Add(float64(sr.Stats.NgramMatches))
 
+					// MaxPendingShardPriority *cannot* be this result's ShardPriority, because
+					// the priority is removed before computing max() and calling sender.Send.
+					// (There may be duplicate priorities, though-- that's fine.) A PendingShard
+					// is one that has not entered this critical section and sent its results.
+					//
+					// Note that there are at least two layers above this implementing streamSearch
+					// or StreamSearch that also take a lock for the entirety of the Send() operation.
+					//
+					// This is to avoid a potential race between shards sending back results
+					// if the priority were removed before sending without a lock:
+					// 1) shard A (pri 1), B (pri 2), C (pri 3) dispatch, pendingPriorities = [1, 2, 3]
+					// 2) C completes and removes itself from the priority list, pP = [1, 2]
+					// 3) B completes, removes itself, computes max, *and sends results* as maxPendingPriority=1,
+					//    indicating that no future results will come from a lower-ordered shard, pP = [1]
+					// 4) A completes, removes itself, computes max, and sends results with maxPP=-Inf, indicating
+					//    that the stream is finished (?)
+					// 5) C finally wakes up, computes max, and sends results with maxPP=-Inf, but with priority=3.
+					mu.Lock()
+					pendingPriorities.remove(s.priority)
+					sr.Stats.MaxPendingShardPriority = pendingPriorities.max()
+					sr.Stats.ShardPriority = s.priority
 					sender.Send(sr)
```

Comment:
I don't quite understand how we're blocking on sending lower priority results if the distance to the max pending priority result grows too large. Isn't it possible in practice that we reach the last shard in the slice while having the first (highest rank) still pending? Won't we just keep sending?

Maybe I misunderstood and we're not actually blocking any sending here, and it's all re-ordered in sourcegraph frontend.

Reply:
As far as I understand it, the intend is not to block but to inform frontend of what is still pending.

Reply:
The blocking doesn't happen on the zoekt side. The frontend delays while it collects results from zoekt.